### PR TITLE
Handle wildcard pointers in SB

### PR DESCRIPTION
**Pull request by carbotaniuman**
This uses an permissive `Unknown` implementation, where a wildcard pointer (and any SRW derived from a wildcard pointer) can access any previously-exposed SB tag. This is missing any meaningful test-cases, and all of the edge-cases have not yet been worked through.

I think there's also some bugs here with differing Unknowns in different ranges and having things behave really weirdly too, alongside some issues with retagging to `SRO` or `Unique`.

### CODE REVIEW

File: src/stacked_borrows.rs [hunk 1]
```diff
@@ -362,6 +377,42 @@ impl<'tcx> Stack {
         Ok(())
     }
 
+    /// Inserts an `SbTag::Unknown` above the specified tag. This represents
+    /// exposing a pointer via an ptrtoint cast.
```
Comment (RalfJung) [inline src/stacked_borrows.rs:381]: Exposing shouldn't change stacks. It doesn't even know which stacks it would possibly have to change.

File: tests/pass/permissive_provenance_sb.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+// random agglomeration test
+use core::cell::UnsafeCell;
+
+fn not_so_innocent(x: &mut UnsafeCell<u32>) {
+    let x_raw: *mut UnsafeCell<u32> = x as *mut UnsafeCell<u32>;
+    x_raw as usize;
+}
+
+fn main() {
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:12]: Please add this as a new testcase to the existing file tests/pass/stacked-borrows/int-to-ptr.rs (and add `-Zmiri-permissive-provenance` to that file).

File: tests/pass/ptr_int_permissive_provenance.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// compile-flags: -Zmiri-permissive-provenance -Zmiri-disable-stacked-borrows
+// compile-flags: -Zmiri-permissive-provenance
```
Comment (RalfJung) [inline tests/pass/ptr_int_permissive_provenance.rs:1]: Please also do the same change in tests/fail/provenance/ptr_int_unexposed.rs

**Review by RalfJung — COMMENTED** [review]
Thanks a lot for working on this. :)
However, I think the logic for this is not quite right yet; I left some comments.

File: src/stacked_borrows/diagnostics.rs [hunk 4]
```diff
@@ -202,6 +202,7 @@ impl AllocHistory {
         alloc_id: AllocId,
         alloc_range: AllocRange,
         error_offset: Size,
+        is_wildcard: bool,
         stack: &Stack,
     ) -> InterpError<'tcx> {
```
Comment (RalfJung) [inline src/stacked_borrows/diagnostics.rs:207]: Feel free to `allow(clippy::too_many_arguments)` to silence clippy

File: src/stacked_borrows.rs [hunk 5]
```diff
@@ -270,20 +274,31 @@ impl Permission {
...
             // We require a permission with the right tag, ensuring U3 and F3.
-            .find_map(
-                |(idx, item)| {
-                    if tag == item.tag && item.perm.grants(access) { Some(idx) } else { None }
-                },
-            )
+            .find_map(|(idx, item)| {
+                // A wildcard tag allows access to all `Unknown` tags.
+                if (tag_is_wildcard && item.tag == SbTag::Unknown || tag == item.tag)
+                    && item.perm.grants(access)
```
Comment (RalfJung) [inline src/stacked_borrows.rs:295]: (Wtf, GH lost the most important comment in my review?!?)

This logic is wrong -- any tag might be in the unknown part, including tags that never got exposed! An unknown thing in the stack represents "here can be literally anything", so the logic should work accordingly.

I still think as a first step we should do the following: keep `Item` and `SbTag` unchanged, and add to `Stack` a `unknown_bottom: bool`. This represents that the stack might go on arbitrarily below the lowest item, but we don't know what it looks like (because it could depend on some angelic choice). Then adjust all the logic accordingly.

Yes, we can do better by exploiting that items can be in a stack at most once, so we can know that certain items can *not* be in the unknown part. But I think we should leave actually exploiting this structure to the future, it is not require for a first useful version of all this.

File: tests/pass/permissive_provenance_sb.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        // [.., Unique(1), SharedRW(2*), Unique(3), ..., SharedRW(4*)]
+        // * denotes exposed
+    
+        // There will be an additional SharedRW(4*) on top of Unique(3) at this
+        // point, if `not_so_innocent` is called.
+    
+        // Cast the integer to the pointer, where confusion may occur about which tag we should use.
+        let x_sharedrw = x_raw1_usize as *mut u32;
+        let x_sharedrw = &*(x_sharedrw as *const UnsafeCell<u32>);
+        let x_sharedrw = &*(x_sharedrw as *const UnsafeCell<u32>);
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:37]: No need to do this twice...

File: tests/pass/permissive_provenance_sb.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        
+        let x_raw1: *mut UnsafeCell<u32> = x_unique1 as *mut UnsafeCell<u32>;
+        let x_raw1_usize = x_raw1 as usize;
+        // [..., Unique(1), SharedRW(2*)]
+        // * denotes exposed
+        
+        let x_unique2: &mut UnsafeCell<u32> = &mut *x_raw1;
+        // [.., Unique(1), SharedRW(2*), Unique(3)]
+        
+        not_so_innocent(x_unique2);
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:27]: Can you make this `if innocent { not_so_innocent }`, add a `innocent: bool` parameter to this function, and then run this entire test twice, for both values of `innocent`?

File: tests/pass/permissive_provenance_sb.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        // this will fail.
+        {
+            *x_unique2 = UnsafeCell::new(10);
+        }
+
+        // Wildcard tags are not magic; popping all exposed tags prevents them from accessing.
+        // Uncomment this to see the next write fail.
+        // {
+        //     c = UnsafeCell::new(45u32);
+        // }
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:55]: We could have this as a failing test, but it'd also be fine to not implement this in the first iteration. In fact I think that would be better -- let's take baby steps.

File: tests/pass/permissive_provenance_sb.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        // There will be an additional SharedRW(4*) on top of Unique(3) at this
+        // point, if `not_so_innocent` is called.
+    
+        // Cast the integer to the pointer, where confusion may occur about which tag we should use.
+        let x_sharedrw = x_raw1_usize as *mut u32;
+        let x_sharedrw = &*(x_sharedrw as *const UnsafeCell<u32>);
+        let x_sharedrw = &*(x_sharedrw as *const UnsafeCell<u32>);
+    
+        // Now try to write to the casted SharedRW. This will use the tag SharedRW(4*)
+        // if not_so_innocent was called, else it will use SharedRW(2*).
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:40]:
```suggestion
        // Now try to write to the casted SharedRW. This must angelically "pick" SharedRW(2*)
        // even if not_so_innocent was called and SharedRW(4*) is further up the stack!
```

File: tests/pass/permissive_provenance_sb.rs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            *x_sharedrw.get() = 12;
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:42]: This write is not there in the original example at https://github.com/rust-lang/unsafe-code-guidelines/issues/273, why is it here?

File: tests/pass/permissive_provenance_sb.rs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        not_so_innocent(x_unique2);
+        // [.., Unique(1), SharedRW(2*), Unique(3), ..., SharedRW(4*)]
+        // * denotes exposed
+    
+        // There will be an additional SharedRW(4*) on top of Unique(3) at this
+        // point, if `not_so_innocent` is called.
+    
+        // Cast the integer to the pointer, where confusion may occur about which tag we should use.
+        let x_sharedrw = x_raw1_usize as *mut u32;
+        let x_sharedrw = &*(x_sharedrw as *const UnsafeCell<u32>);
```
Comment (RalfJung) [inline tests/pass/permissive_provenance_sb.rs:36]: FWIW all this `UnsafeCell` stuff should be unnecessary now since we have raw ptr tagging.

**Comment by RalfJung** [comment]
> This is missing any meaningful test-cases, and all of the edge-cases have not yet been worked through.

I think enabling stacked borrows for the existing permissive provenance tests, and the existing `stacked-borrows/int-to-ptr.rs` is a good baseline. The test you added that is based on https://github.com/rust-lang/unsafe-code-guidelines/issues/273 is great; I think it is extremely meaningful -- but we don't need `UnsafeCell` any more and your version also added an extra write that I think makes it the wrong test; here's what we should do instead I think:
```rust
#![feature(strict_provenance)]
use std::ptr;

/// Ensure that we do not just pick the topmost possible item on int2ptr casts.
fn example(variant: bool) { unsafe {
    fn not_so_innocent(x: &mut u32) -> usize {
        let x_raw = x as *mut u32;
        x_raw.expose_addr()
    }

    let mut c = 42u32;

    let x_unique1 = &mut c;
    // [..., Unique(1)]
    
    let x_raw2 = x_unique1 as *mut u32;
    let x_raw2_addr = x_raw2.expose_addr();
    // [..., Unique(1), SharedRW(2)]
    
    let x_unique3 = &mut *x_raw2;
    // [.., Unique(1), SharedRW(2), Unique(3)]
    
    assert_eq!(not_so_innocent(x_unique3), x_raw2_addr);
    // [.., Unique(1), SharedRW(2), Unique(3), ..., SharedRW(4)]
    
    // Do an int2ptr cast. This can pick tag 2 or 4 (the two previously exposed tags).
    // 4 is the "obvious" choice (topmost tag, what we used to do with untagged pointers).
    // And indeed if `variant == true` it is the only possible choice.
    // But if `variant == false` then 2 is the only possible choice!
    let x_wildcard = ptr::from_exposed_addr_mut::<i32>(x_raw2_addr);
    
    if variant {
        // If we picked 2, this will invalidate 3.
        *x_wildcard = 10;
        // Now we use 3. Only possible if above we picked 4.
        *x_unique3 = 12;
    } else {
        // This invalidates tag 4.
        *x_unique3 = 10;
        // Now try to write with the "guessed" tag; it must be 2.
        *x_wildcard = 12;
    }
} }

fn main() {
    example(false);
    example(true);
}
```

In terms of failing tests, the important ones are the ones in `fail/stacked_borrows` -- they should all keep failing. You could try running `MIRIFLAGS="-Zmiri-permissive-provenance" ./miri test fail/stacked` to see how they are doing.

In terms of further passing tests, here's another very simple one:
```rust
fn main() { unsafe {
    let root = &mut 42;
    let root_raw = root as *mut i32;
    let addr1 = root_raw as usize;
    let child = &mut *root_raw;
    let child_raw = child as *mut i32;
    let addr2 = child_raw as usize;
    assert_eq!(addr1, addr2);
    // First use child.
    *(addr2 as *mut i32) -= 2; // picks child_raw
    *child -= 2;
    // Then use root.
    *(addr1 as *mut i32) += 2; // picks root_raw
    *root += 2;
    // Value should be unchanged.
    assert_eq!(*root, 42);
} }
```

File: src/stacked_borrows.rs (hunk 5 above)
Comment (carbotaniuman) [reply src/stacked_borrows.rs:295]: > This logic is wrong -- any tag might be in the unknown part, including tags that never got exposed!

How could this work again? Sorry if it was discussed previously, it's been a while. Wouldn't that mean that any tags above the Unknown don't matter, as even if they got popped, we wouldn't know if they were still in the Unknown?

Comment (RalfJung) [reply src/stacked_borrows.rs:295]: Imagine the stack is (ignoring permissions)
```
SharedRW(4)
Unique(3)
SharedRW(2)
Unique(1)
```
with 2, 4 exposed. Now a write is done with a wildcard pointer. There are two possible resulting stacks:
```
SharedRW(4)
Unique(3)
SharedRW(2)
Unique(1)
```
or
```
SharedRW(2)
Unique(1)
```
We can summarize both stacks by storing
```
UNKNOWN
Unique(1)
```
Notice how "UNKNOWN" might contain a `Unique(3)`, even though that tag is not exposed!

As a further simplification, I think if we have an UNKNOWN we should also "forget" everything below that UNKNOWN, so the resulting stack here would just be
```
UNKNOWN
```
Yes, this will miss some bugs, but crucially it still finds the following bug, and that is good enough for now:
```rust
let root = &mut 42;
let addr = root as *mut i32 as usize;
let root2 = &mut *(addr as *mut i32);
let raw2 = root2 as *mut i32;
*root2 = 0;
let _val = *raw2; // UB!
```
I'd rather land something basic soon than take a long time to build a more precise but more complicated analysis. :)

Comment (carbotaniuman) [reply src/stacked_borrows.rs:295]: I guess my main question is once we execute `*root2 = 0;` and the tag associated with `raw2` is popped, how does the code know it's actually not down in the `UNKNOWN` section of the stack?

Comment (RalfJung) [reply src/stacked_borrows.rs:295]: Oh, that is a good point.

I think we can handle this by exploiting that tags are monotonically increasing: the stack after `let addr` looks like
```
SharedRW(2) // tag of the `root as *mut _`
Unique(1) // tag of root
```
when we access this with the wildcard pointer, we notice 2 is exposed, so we replace the stack by `UNKNOWN(next_ptr_id)`, meaning "an unknown set of items where all tags are strictly less than `next_ptr_id`". This is trivially correct because *all* currently existing tags are strictly less than `next_ptr_id`.

`root2` and `raw2` will get a new tag that is *not* strictly less than `next_ptr_id`, so we know it cannot be in the unknown part.

Comment (carbotaniuman) [reply src/stacked_borrows.rs:295]: Hmm one thing about this is that it doesn't fix https://github.com/rust-lang/unsafe-code-guidelines/issues/273, we are effectively doing the same thing as choosing the top-most exposed when it comes to reborrows.

Given the other numerous missed bugs with this that's probably fine, but I just wanted to make sure I'm getting the reborrow logic right for wildcard ponters (blow away everything below the exposed tag and replace with Unknown, then use the Unknown as the granting tag).

Comment (RalfJung) [reply src/stacked_borrows.rs:295]: > we are effectively doing the same thing as choosing the top-most exposed when it comes to reborrows.

Is "this" your proposal or mine? Under mine, we are *not* doing that. That is the entire point.

File: src/stacked_borrows.rs [hunk 12]
```diff
@@ -80,16 +78,33 @@ impl fmt::Debug for Item {
     }
 }
 
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub enum ItemOrUnknown {
+    Item(Item),
+    Unknown(PtrId),
```
Comment (RalfJung) [inline src/stacked_borrows.rs:84]: Sine you have `unknown_bottom`, what is this for?

File: src/stacked_borrows.rs [hunk 13]
```diff
@@ -80,16 +78,33 @@ impl fmt::Debug for Item {
...
     /// Invariants:
     /// * Above a `SharedReadOnly` there can only be more `SharedReadOnly`.
     /// * Except for `Untagged`, no tag occurs in the stack more than once.
-    borrows: Vec<Item>,
-    /// Tags that should be treated as wildcards.
-    treated_as_wildcard: FxHashSet<SbTag>,
+    borrows: Vec<ItemOrUnknown>,
+    /// Whether the borrow stack extends for an unknown amount of items at the bottom
+    /// of the stack, whose PtrId of it's SbTag is less than this.
+    // TODO: reword this
```
Comment (RalfJung) [inline src/stacked_borrows.rs:106]:
```suggestion
    /// If this is `Some(id)`, then the actual current stack is unknown. What we do know
    /// is that `borrows` are at the top of the stack, and below it are arbitrarily many items
    /// whose `tag` is either `Untagged` or strictly less than `id`.
```

File: src/stacked_borrows.rs [hunk 14]
```diff
@@ -106,6 +121,8 @@ pub struct Stacks {
 pub struct GlobalStateInner {
     /// Next unused pointer ID (tag).
     next_ptr_id: PtrId,
+    /// The tags that have been exposed
+    exposed_tags: FxHashSet<SbTag>,
```
Comment (RalfJung) [inline src/stacked_borrows.rs:125]: I think you can move this into `Stacks`, since it is per-allocation state.

File: tests/pass/ptr_int_permissive_provenance.rs (hunk 3 above)
Comment (RalfJung) [reply tests/pass/ptr_int_permissive_provenance.rs:1]: And also in tests/fail/provenance/permissive_provenance_transmute.rs

File: src/stacked_borrows.rs (hunk 12 above)
Comment (RalfJung) [reply src/stacked_borrows.rs:84]: Using a `Vec<ItemOrUnknown>` will probably make the items in that vec bigger than they used to be, which is a problem since Stacked Borrows already leads to enormous memory consumption for big stacks. We should really avoid growing the type of elements of that `Vec`.

File: src/stacked_borrows.rs [hunk 15]
```diff
@@ -283,32 +289,72 @@ impl Permission {
...
             .rev() // search top-to-bottom
             // Return permission of first item that grants access.
             // We require a permission with the right tag, ensuring U3 and F3.
-            .find_map(
-                |(idx, item)| {
-                    if tag == item.tag && item.perm.grants(access) { Some(idx) } else { None }
+            .find_map(|(idx, item)| {
+                match tag {
+                    Some(tag) if tag == item.tag && item.perm.grants(access) => Some(idx),
+                    None if exposed_tags.contains(&item.tag) => Some(idx),
```
Comment (RalfJung) [inline src/stacked_borrows.rs:310]: I think this will become cleaner if you handle the `None` tag first in a separate branch all the way at the top of the function. We don't need the `idx` in that case anyway, I think?

Also, the logic here is wrong -- we still need to check `item.perm.grants(access)`. We should probably have a testcase for that: if only SRO tags have been exposed, then a write access with a wildcard pointer should fail.

File: src/stacked_borrows.rs [hunk 16]
```diff
@@ -380,58 +426,66 @@ impl<'tcx> Stack {
...
         access: AccessKind,
-        tag: SbTag,
+        tag: Option<SbTag>,
         (alloc_id, alloc_range, offset): (AllocId, AllocRange, Size), // just for debug printing and error messages
         global: &mut GlobalStateInner,
         current_span: &mut CurrentSpan<'_, '_, 'tcx>,
         alloc_history: &mut AllocHistory,
+        exposed_tags: &FxHashSet<SbTag>,
     ) -> InterpResult<'tcx> {
         // Two main steps: Find granting item, remove incompatible items above.
```
Comment (RalfJung) [inline src/stacked_borrows.rs:436]: In fact, maybe this is the right place to separately handle `tag == None`? I am not sure if there is much code that can be shared between those cases. We might want a separate `access_wildcard` function and put all the wildcard logic in there.

File: src/machine.rs [hunk 17]
```diff
@@ -695,8 +695,24 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
...
+            Tag::Concrete(ConcreteTag { alloc_id, sb }) => {
+                intptrcast::GlobalStateInner::expose_addr(ecx, alloc_id);
+
+                let (size, _) =
+                    ecx.get_alloc_size_and_align(alloc_id, AllocCheck::MaybeDead).unwrap();
+
+                // Function pointers and dead objects don't have an alloc_extra so we ignore them.
+                if let Ok(alloc_extra) = ecx.get_alloc_extra(alloc_id) {
+                    if let Some(stacked_borrows) = &alloc_extra.stacked_borrows {
+                        stacked_borrows.ptr_exposed(
```
Comment (RalfJung) [inline src/machine.rs:707]: `expose_tag` would be a better name for this function IMO.

Also the function doesn't need to take an `alloc_range` I think.

File: src/machine.rs [hunk 18]
```diff
@@ -695,8 +695,24 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
...
-            Tag::Concrete(concrete) =>
-                intptrcast::GlobalStateInner::expose_addr(ecx, concrete.alloc_id),
+            Tag::Concrete(ConcreteTag { alloc_id, sb }) => {
+                intptrcast::GlobalStateInner::expose_addr(ecx, alloc_id);
+
+                let (size, _) =
+                    ecx.get_alloc_size_and_align(alloc_id, AllocCheck::MaybeDead).unwrap();
+
+                // Function pointers and dead objects don't have an alloc_extra so we ignore them.
+                if let Ok(alloc_extra) = ecx.get_alloc_extra(alloc_id) {
```
Comment (RalfJung) [inline src/machine.rs:705]: This "catches" an `InterpError`. Please don't do that; use `?` instead.

Comment (carbotaniuman) [reply src/machine.rs:705]: I just tested this again, and the comment is still accurate. I need to filter out function pointers somehow, or else I get
```
Undefined Behavior: accessing alloc127 which contains a function
```
with the code being something like
```
action.sa_sigaction = signal_handler as sighandler_t
```

Comment (RalfJung) [reply src/machine.rs:705]: Ah, right. The pointer might also dangle and this allocation no longer exist.

Catching `InterpError` is not the right solution, though. I don't have a much better idea right now... maybe putting the 'exposed' map into `AllocExtra` is not the right call after all. We can put it into `GlobalStateInner` for now and optimize this later if need be.

**Comment by RalfJung** [comment]
All right I think this is now a pretty reasonable first implementation for permissive provenance in Stacked Borrows. :) It can even detect some incorrect uses of mutable and shared references derived from wildcard pointers. It probably still misses some UB that the "Untagged" approach would have caught, but on the flip side it is actually *correct* and doesn't suffer from https://github.com/rust-lang/unsafe-code-guidelines/issues/273. Also it misses no UB when you never use a wildcard pointer (i.e., a pointer obtained via `from_exposed_addr` or an int2ptr cast).

@saethlin could you check if the changes to your diagnostics code make sense? It is mostly doing nothing now when encountering a wildcard pointer, which can probably be improved -- in a follow-up PR, if anyone wants to work on that. :)

File: src/stacked_borrows.rs [hunk 19]
```diff
@@ -104,6 +131,13 @@ pub struct Stack {
     /// * Above a `SharedReadOnly` there can only be more `SharedReadOnly`.
     /// * Except for `Untagged`, no tag occurs in the stack more than once.
     borrows: Vec<Item>,
+    /// If this is `Some(id)`, then the actual current stack is unknown. THis can happen when
```
Comment (saethlin) [inline src/stacked_borrows.rs:134]:
```suggestion
    /// If this is `Some(id)`, then the actual current stack is unknown. This can happen when
```

File: tests/fail/stacked_borrows/exposed_only_ro.rs [hunk 20]
```diff
@@ -0,0 +1,12 @@
+// compile-flags: -Zmiri-permissive-provenance
+#![feature(strict_provenance)]
+
+// If we have only exposed read-only pointers, doing a write through a wildcard ptr should fail.
```
Comment (saethlin) [inline tests/fail/stacked_borrows/exposed_only_ro.rs:4]: This comment doesn't line up with the diagnostic. If this comment is correct, the diagnostic should say something like "but no tags for this location with write permission were exposed". Because a user could complain "This error message says no exposed tags are valid, but look I exposed a tag right here by calling `expose_addr` and I never invalidated it".

File: tests/fail/stacked_borrows/illegal_read_despite_exposed2.rs [hunk 21]
```diff
@@ -0,0 +1,18 @@
+// compile-flags: -Zmiri-permissive-provenance
+
+fn main() {
+    unsafe {
+        let root = &mut 42;
+        let addr = root as *mut i32 as usize;
+        let exposed_ptr = addr as *mut i32;
+        // From the exposed ptr, we get a new unique ptr.
+        let root2 = &mut *exposed_ptr;
+        // let _fool = root2 as *mut _; // this would [fool] us, since SRW(N+1) remains on the stack
```
Comment (saethlin) [inline tests/fail/stacked_borrows/illegal_read_despite_exposed2.rs:10]: Why is this `[fool]` not just `fool`?

File: tests/fail/stacked_borrows/illegal_read_despite_exposed1.rs [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+        let addr = root as *mut i32 as usize;
+        let exposed_ptr = addr as *mut i32;
+        // From the exposed ptr, we get a new unique ptr.
+        let root2 = &mut *exposed_ptr;
+        let _fool = root2 as *mut _; // this would have fooled the old untagged pointer logic
+        // Stack: Unknown(<N), Unique(N), SRW(N+1)
+        // And we test that it has uniqueness by doing a conflicting write.
+        *exposed_ptr = 0;
+        // Stack: Unknown(<N)
+        let _val = *root2; //~ ERROR: borrow stack
```
Comment (saethlin) [inline tests/fail/stacked_borrows/illegal_read_despite_exposed1.rs:15]: This collapsing behavior seems very important and is worth having a note for, like we do for tag invalidation. That's certainly something I can do after this PR.

**Comment by saethlin** [comment]
The changes to the diagnostic _code_ mostly make sense, but for me the question is whether these diagnostics help a user make sense of errors they've never seen before. I'll run this over a bunch of crates over the next few days and probably want to upgrade the diagnostics as I indicated.

But in case this is what you were asking, I don't see anything that's bad enough to warrant immediate concern, on account of how this is all behind a non-default flag.

File: tests/fail/stacked_borrows/exposed_only_ro.rs (hunk 20 above)
Comment (RalfJung) [reply tests/fail/stacked_borrows/exposed_only_ro.rs:4]: Ah, fair. We can probably make a more targeted diagnostic but for now I'll just adjust the error text.

File: tests/fail/stacked_borrows/illegal_read_despite_exposed1.rs (hunk 22 above)
Comment (RalfJung) [reply tests/fail/stacked_borrows/illegal_read_despite_exposed1.rs:15]: Note that collapsing is *not* part of the spec -- it is part of the approximation Miri does because the exact spec (as documented [here](https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html)) cannot be reasonably implemented. So, in the future, if we find a better way to approximate this, we might change this.

I guess what I am saying is -- this might be even harder to communicate than the regular Stacked Borrows rules, since these are derived rules that are computed from the actual rules.

**Comment by RalfJung** [comment]
Thanks! Yes that is basically what I was asking.

Though I do plan to make this flag on-by-default very soon. However, all the new cases can only arise in code that actually uses `from_exposed_addr`, and meanwhile this entirely *removes* all the confusion caused by `Untagged`, so I think it is still a net win.

**Comment by RalfJung** [comment]
Thanks a lot @carbotaniuman for laying the foundation here! With this patch, `-Zmiri-permissive-provenance` should be production-ready (but with room for improvement in its diagnostics). I am very excited about getting rid of `Untagged` soon. :)

@bors r+
